### PR TITLE
Add toExprOfTuple in scala.quoted

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by milessabin** [comment]
LGTM!

I think we also need something to construct a tuple _type_ given a `Seq[Type]`.

**Comment by nicolasstucki** [comment]
Still need to add at least 23 test

**Comment by nicolasstucki** [comment]
What would be the signature of the `Seq[Type]`?

**Comment by milessabin** [comment]
> What would be the signature of the `Seq[Type]`?

How about,

```scala
def tupleType(tpes: Seq[reflect.Type]): reflect.Type
```

Or should that be `quoted.Type`? I have to be honest, I don't yet have a clear mental model of when I should be working with quoted vs tasty.reflect types.

**Comment by milessabin** [comment]
The context is, I want to be able to generate a trait implementation of the form,

```scala
trait Foo {
  type ElemTypes = <tuple type here>
  val Elems: ElemTypes = <tuple term here>
}
```
where the tuple elements are typed tasty trees (from an `annots` call) and the tuple element types are the tasty types recovered from those trees.

I'm open to changing the approach so long as I can produce a term of that form from whatever I get back from `annots`.

**Comment by nicolasstucki** [comment]
That type for tuple is completely unrelated to this PR. I will add work on It in a different PR.

**Comment by milessabin** [comment]
> That type for tuple is completely unrelated to this PR. I will add work on It in a different PR.

Awesome ... thanks. And let me know if there's anything I can do to help here.

**Comment by nicolasstucki** [comment]
> > That type for tuple is completely unrelated to this PR. I will add work on It in a different PR.
> 
> Awesome ... thanks. And let me know if there's anything I can do to help here.

Could you make minimal full example/test where the only thing missing is the type of the tuple?

**Comment by anatoliykmetyuk** [comment]
Possibly related: #7047 though it lifts tuples, not sequences, and is currently blocked.

**Comment by milessabin** [comment]
@nicolasstucki I'm not sure if that's exactly what you meant by "Could you make minimal full example/test where the only thing missing is the type of the tuple?", but the test I've just added _doesn't_ compile if you uncomment the type annotations, failing with,

```
-- Error: tests/run/quote-toExprOfTuple/Test_2.scala:2:19 ----------------------
2 |  val t2 = Macro.t2(23, "foo")
  |           ^^^^^^^^^^^^^^^^^^^
  |Exception occurred while executing macro expansion.
  |scala.tasty.reflect.ExprCastError: Expr: {
  |  Tuple2.apply[Any, Any](
  |    {
  |      {
  |        23
  |      }
  |    }
  |  , 
  |    {
  |      {
  |        "foo"
  |      }
  |    }
  |  )
  |}
  |did not conform to type: (Int, String)
  |
  |     at dotty.tools.dotc.tastyreflect.ReflectionCompilerInterface.QuotedExpr_cast(ReflectionCompilerInterface.scala:1767)
  |     at dotty.tools.dotc.tastyreflect.ReflectionCompilerInterface.QuotedExpr_cast(ReflectionCompilerInterface.scala:1759)
  |     at scala.tasty.reflect.QuotedOps$QuotedExprAPI.cast(QuotedOps.scala:13)
  |     at Macro$.impl2(Macro_1.scala:12)
  |
  | This location is in code that was inlined at Test_2.scala:2
one error found
[error] Nonzero exit code returned from runner: 1
[error] (dotty-compiler-bootstrapped / Compile / runMain) Nonzero exit code returned from runner: 1
[error] Total time: 3 s, completed 15-Aug-2019 16:05:29
sbt:dotty>
```

File: tests/run/quote-toExprOfTuple/Macro_1.scala [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import scala.quoted._
+
+object Macro {
```
Comment (nicolasstucki) [inline tests/run/quote-toExprOfTuple/Macro_1.scala]: This test should be in `tests/run-macros`

Comment (nicolasstucki) [reply tests/run/quote-toExprOfTuple/Macro_1.scala]: That fixed it. It was probably a bad interaction with the non bootstrapped compiler.

**Comment by milessabin** [comment]
I'm a bit confused ... the test I added _didn't_ compile with the type annotation uncommented, which looked wrong to me.

Was that only the case because the test was being compiled with the wrong flags? If that's so then the test should have the type annotations restored, surely?

**Comment by nicolasstucki** [comment]
I will have a second look at it

**Comment by nicolasstucki** [comment]
Followed by #7076

**Pull request merged**